### Word Embedding Techniques using Embedding Layer in Keras

In [ ]:
### Libraries USed Tensorflow> 2.0  and keras

In [2]:
!pip install tensorflow-gpu # No need to install this cell explicitly.

  Using cached tensorflow-gpu-2.12.0.tar.gz (2.6 kB)
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


## Importing Tensorflow Library

In [3]:
import tensorflow as tf

print(tf.__version__) # checking the version of tensorflow

2.18.0


## Importing OHE from tf

In [4]:
##tensorflow >2.0
from tensorflow.keras.preprocessing.text import one_hot

## Declaring Sentences

In [5]:
### sentences
sent=[  'the glass of milk',
     'the glass of juice',
     'the cup of tea',
    'I am a good boy',
     'I am a good developer',
     'understand the meaning of words',
     'your videos are good']

In [6]:
sent

['the glass of milk',
 'the glass of juice',
 'the cup of tea',
 'I am a good boy',
 'I am a good developer',
 'understand the meaning of words',
 'your videos are good']

## Explicitly specifying Vocabulary

In [15]:
### Vocabulary size - More the size more the processing

# Its a Hyperparameter. More the vocabulary size, we will get good feature representation.

voc_size=500

## One Hot Representation

In [13]:
onehot_repr = [one_hot(sentence,voc_size)for sentence in sent]

print(onehot_repr)

[[1220, 8459, 2709, 7322], [1220, 8459, 2709, 4829], [1220, 1617, 2709, 3836], [5, 7174, 8535, 6742, 3524], [5, 7174, 8535, 6742, 3097], [5898, 1220, 5065, 2709, 8257], [3628, 5702, 7515, 6742]]


### Understanding above output

As we specified the vocabulary size explicitly as 10,000. An array of 10,000 indexes gets generated and once we pass the each sentence to the `one_hot(sentence, voc_size)` iterating over the sentences list, then it will return the output of the indexes where that particular word is present in the given vocabulary.

For example: [1220, 8459, 2709, 7322]

Here 1220 means for the word 'the' index is 1220 in the vocabulary.

## Word Embedding Represntation

In [16]:
from tensorflow.keras.layers import Embedding
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential

In [17]:
import numpy as np

## Padding

We need the size of all the sentences to be equal while training the ANN.

Hence, we will specify the maximum Sentence length or more here we are keeping as 8, and then we will use help of **Padding** where it will add zeroes at the front or at the end based on "pre" or "post" padding specified.

In [18]:
## pre padding
sent_length=8

embedded_docs=pad_sequences(onehot_repr,padding='pre',maxlen=sent_length)

print(embedded_docs)

[[   0    0    0    0 1220 8459 2709 7322]
 [   0    0    0    0 1220 8459 2709 4829]
 [   0    0    0    0 1220 1617 2709 3836]
 [   0    0    0    5 7174 8535 6742 3524]
 [   0    0    0    5 7174 8535 6742 3097]
 [   0    0    0 5898 1220 5065 2709 8257]
 [   0    0    0    0 3628 5702 7515 6742]]


## Declaring Feature Dimensions

In [22]:
## 10 feature dimesnions

dim=10


In [23]:
model=Sequential()

# Embedding layer works similar to Word2Vec
model.add(Embedding(voc_size, # Vocabulary size
                    10, # Dimension lenghth - Features required for each vector (means every vector represented in 10 dimensions)
                    input_length=sent_length)) # Sentences lenght

model.compile('adam','mse') # Compiling with Adam optimizer and MSE loss function

In [24]:
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_2 (Embedding)              │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

## Predicting

In [27]:
model.predict(embedded_docs[0])

1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 858ms/step


array([[-0.01417743,  0.02342245, -0.00028545, -0.04773829,  0.0269583 ,
        -0.03326305,  0.04963494,  0.0115789 , -0.03794973, -0.01383566],
       [-0.01417743,  0.02342245, -0.00028545, -0.04773829,  0.0269583 ,
        -0.03326305,  0.04963494,  0.0115789 , -0.03794973, -0.01383566],
       [-0.01417743,  0.02342245, -0.00028545, -0.04773829,  0.0269583 ,
        -0.03326305,  0.04963494,  0.0115789 , -0.03794973, -0.01383566],
       [-0.01417743,  0.02342245, -0.00028545, -0.04773829,  0.0269583 ,
        -0.03326305,  0.04963494,  0.0115789 , -0.03794973, -0.01383566],
       [ 0.03374443, -0.04046713,  0.01686255,  0.03337267,  0.03554653,
         0.01253268,  0.01704322, -0.02261758, -0.0099019 , -0.0373638 ],
       [ 0.03374443, -0.04046713,  0.01686255,  0.03337267,  0.03554653,
         0.01253268,  0.01704322, -0.02261758, -0.0099019 , -0.0373638 ],
       [ 0.03374443, -0.04046713,  0.01686255,  0.03337267,  0.03554653,
         0.01253268,  0.01704322, -0.02261758

In [29]:
embedded_docs[0] # After padding first sentence - the glass of milk

array([   0,    0,    0,    0, 1220, 8459, 2709, 7322], dtype=int32)

In [32]:
model.predict(embedded_docs[1])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 32ms/step


array([[-0.01417743,  0.02342245, -0.00028545, -0.04773829,  0.0269583 ,
        -0.03326305,  0.04963494,  0.0115789 , -0.03794973, -0.01383566],
       [-0.01417743,  0.02342245, -0.00028545, -0.04773829,  0.0269583 ,
        -0.03326305,  0.04963494,  0.0115789 , -0.03794973, -0.01383566],
       [-0.01417743,  0.02342245, -0.00028545, -0.04773829,  0.0269583 ,
        -0.03326305,  0.04963494,  0.0115789 , -0.03794973, -0.01383566],
       [-0.01417743,  0.02342245, -0.00028545, -0.04773829,  0.0269583 ,
        -0.03326305,  0.04963494,  0.0115789 , -0.03794973, -0.01383566],
       [ 0.03374443, -0.04046713,  0.01686255,  0.03337267,  0.03554653,
         0.01253268,  0.01704322, -0.02261758, -0.0099019 , -0.0373638 ],
       [ 0.03374443, -0.04046713,  0.01686255,  0.03337267,  0.03554653,
         0.01253268,  0.01704322, -0.02261758, -0.0099019 , -0.0373638 ],
       [ 0.03374443, -0.04046713,  0.01686255,  0.03337267,  0.03554653,
         0.01253268,  0.01704322, -0.02261758

In [ ]:
#Assignment

sent=["The world is a better place",
      "Marvel series is my favourite movie",
      "I like DC movies",
      "the cat is eating the food",
      "Tom and Jerry is my favourite movie",
      "Python is my favourite programming language"
      ]